br/pkg/restore/client.go

```diff
@@ -2047,41 +2077,128 @@ func (rc *Client) RestoreMetaKVFiles(
 	return nil
 }
 
-// RestoreMetaKVFile tries to restore a file about meta kv-event from stream-backup.
-func (rc *Client) RestoreMetaKVFile(
+func (rc *Client) RestoreMetaKVFilesWithBatchMethod(
 	ctx context.Context,
-	file *backuppb.DataFileInfo,
-	sr *stream.SchemasReplace,
-) (uint64, uint64, error) {
+	files []*backuppb.DataFileInfo,
+	schemasReplace *stream.SchemasReplace,
+	updateStats func(kvCount uint64, size uint64),
+	progressInc func(),
+	restoreBatch func(
+		ctx context.Context,
+		files []*backuppb.DataFileInfo,
+		schemasReplace *stream.SchemasReplace,
+		updateStats func(kvCount uint64, size uint64),
+		progressInc func(),
+	) error,
+) error {
 	var (
-		kvCount uint64
-		size    uint64
+		rangeMin uint64
+		rangeMax uint64
+		idx      int
 	)
-	log.Info("restore meta kv events", zap.String("file", file.Path),
-		zap.String("cf", file.Cf), zap.Int64("kv-count", file.NumberOfEntries),
-		zap.Uint64("min-ts", file.MinTs), zap.Uint64("max-ts", file.MaxTs))
+	for i, f := range files {
+		if i == 0 {
+			idx = i
+			rangeMax = f.MaxTs
+			rangeMin = f.MinTs
+		} else {
+			if f.MinTs <= rangeMax {
+				rangeMin = mathutil.Min(rangeMin, f.MinTs)
+				rangeMax = mathutil.Max(rangeMax, f.MaxTs)
+			} else {
+				err := restoreBatch(ctx, files[idx:i], schemasReplace, updateStats, progressInc)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				idx = i
+				rangeMin = f.MinTs
+				rangeMax = f.MaxTs
+			}
+		}
+
+		if i == len(files)-1 {
```

Comment:
should it be `if idx < len(files-1) {` 🤔

Reply:
It represents that the loop reach the last file, we need to restoreBatch() the rest file at the end.